src/data/nat/modeq.lean

```diff
@@ -154,6 +168,79 @@ lemma le_of_lt_add (h1 : a ≡ b [MOD m]) (h2 : a < b + m) : a ≤ b :=
 lemma add_le_of_lt (h1 : a ≡ b [MOD m]) (h2 : a < b) : a + m ≤ b :=
 le_of_lt_add (add_modeq_right.trans h1) (add_lt_add_right h2 m)
 
+lemma dvd_iff_of_modeq_of_dvd {a b d m : ℕ} (h : a ≡ b [MOD m]) (hdm : d ∣ m) :
+  d ∣ a ↔ d ∣ b :=
+begin
+  simp only [←modeq_zero_iff_dvd],
+  have habd := modeq.modeq_of_dvd hdm h,
+  exact ⟨(modeq.symm habd).trans, modeq.trans habd⟩,
+end
+
+lemma gcd_eq_of_modeq {a b m : ℕ} (h : a ≡ b [MOD m]) : gcd a m = gcd b m :=
+begin
+  have h1 := gcd_dvd_right a m,
+  have h2 := gcd_dvd_right b m,
+  exact dvd_antisymm
+    (dvd_gcd ((dvd_iff_of_modeq_of_dvd h h1).mp (gcd_dvd_left a m)) h1)
+    (dvd_gcd ((dvd_iff_of_modeq_of_dvd h h2).mpr (gcd_dvd_left b m)) h2),
+end
+
+lemma eq_of_modeq_of_abs_lt {a b m : ℕ} (h : a ≡ b [MOD m]) (h2 : | (b:ℤ) - a | < m) : a = b :=
+begin
+  apply int.coe_nat_inj,
+  rw [eq_comm, ←sub_eq_zero],
+  exact int.eq_zero_of_abs_lt_dvd (modeq_iff_dvd.mp h) h2,
+end
+
+/-- To cancel a common factor `c` from a `modeq` we must divide the modulus `m` by `gcd m c` -/
+lemma modeq_cancel_left_div_gcd {a b c m : ℕ} (hm : 0 < m) (h : c * a ≡ c * b [MOD m]) :
+  a ≡ b [MOD m / gcd m c] :=
+begin
+  set d := (gcd m c),
+  have hmd := gcd_dvd_left m c,
+  have hcd := gcd_dvd_right m c,
+  rw modeq_iff_dvd,
+  have h1 : ((m:ℤ)/↑d) ∣ (↑c/↑d) * (↑b - ↑a),
+  { rw [mul_comm, ←int.mul_div_assoc (↑b - ↑a) (int.coe_nat_dvd.mpr hcd), mul_comm],
+    refine int.div_dvd_div (int.coe_nat_dvd.mpr hmd) _,
+    rw mul_sub,
+    exact modeq_iff_dvd.mp h },
+  have h2 : int.gcd ((m/d):ℤ) (c/d) = 1 :=
+    by simp only [←int.coe_nat_div, int.coe_nat_gcd (m / d) (c / d), gcd_div hmd hcd,
+        nat.div_self (gcd_pos_of_pos_left c hm)],
+  exact int.dvd_of_dvd_mul_right_of_gcd_one h1 h2,
+end
```

Comment:
There are a number of coercions that don't need to be written. I also changed the organization of the argument a little using `refine` and `show`.  The `set` is a higher-level tactic than `let`, so I switched it to that, too.
```suggestion
begin
  let d := gcd m c,
  have hmd := gcd_dvd_left m c,
  have hcd := gcd_dvd_right m c,
  rw modeq_iff_dvd,
  refine int.dvd_of_dvd_mul_right_of_gcd_one _ _,
  show (m/d : ℤ) ∣ (c/d) * (b - a),
  { rw [mul_comm, ←int.mul_div_assoc (b - a) (int.coe_nat_dvd.mpr hcd), mul_comm],
    apply int.div_dvd_div (int.coe_nat_dvd.mpr hmd),
    rw mul_sub,
    exact modeq_iff_dvd.mp h },
  show int.gcd (m/d) (c/d) = 1,
  { simp only [←int.coe_nat_div, int.coe_nat_gcd (m / d) (c / d), gcd_div hmd hcd,
      nat.div_self (gcd_pos_of_pos_left c hm)], }
end
```

Reply:
That's much nicer, thanks